datasource/src/main/scala/quasar/physical/mongo/decoder.scala

```diff
@@ -0,0 +1,154 @@
+/*
+ * Copyright 2014–2018 SlamData Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package quasar.physical.mongo
+
+import slamdata.Predef._
+import qdata._
+import qdata.QType._
+import quasar.contrib.std.errorImpossible
+
+import java.time._
+import java.util.{Map, Iterator}
+
+import eu.timepit.refined.auto._
+import org.bson._
+import spire.math.Real
+
+object decoder {
+  @SuppressWarnings(Array("org.wartremover.warts.Var"))
+  class BsonDocumentCursor(document: BsonDocument) {
+
+    val iterator: Iterator[Map.Entry[String, BsonValue]] =
+      document.entrySet().iterator()
+
+    var entry: Option[Map.Entry[String, BsonValue]] = None
+
+    def hasNext(): Boolean = {
```

Comment:
This is kinda interesting: the code checks if object or array `hasNext` in qdata and then directly takes values and keys, calling `step` in the last command of while-loop. It's convenient for lists I think, but iterator needs to call `next` at first and then do nothing.

Reply:
Aye, that's a good point. Maybe instead of `hasNext` this should be named `isEmpty`?

Also, using
```scala
if (iterator.hasNext()) {
  ...
} else {
  false
}
```
is preferable to the try/catch block.